object/slicer/slicer.go

```diff
@@ -166,40 +199,76 @@ func (x *Slicer) Slice(data io.Reader, attributes ...object.Attribute) (oid.ID,
 	return rootID, nil
 }
 
-// InitPayloadStream works similar to Slice but provides PayloadWriter allowing
-// the caller to write data himself.
-func (x *Slicer) InitPayloadStream(attributes ...object.Attribute) (*PayloadWriter, error) {
+// headerData extract required fields from header, otherwise throw the error.
+func headerData(header object.Object) (cid.ID, user.ID, error) {
+	containerID, isSet := header.ContainerID()
+	if !isSet {
+		return cid.ID{}, user.ID{}, fmt.Errorf("container-id: %w", ErrIncompleteHeader)
+	}
+
+	owner := header.OwnerID()
+	if owner == nil {
+		return cid.ID{}, user.ID{}, fmt.Errorf("owner: %w", ErrIncompleteHeader)
+	}
+
+	return containerID, *owner, nil
+}
+
+func initPayloadStream(ctx context.Context, ow ObjectWriter, header object.Object, signer neofscrypto.Signer, opts Options) (*PayloadWriter, error) {
+	containerID, owner, err := headerData(header)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.sessionToken != nil {
+		header.SetSessionToken(opts.sessionToken)
+		// session issuer is a container owner.
+		issuer := opts.sessionToken.Issuer()
+		owner = issuer
+		header.SetOwnerID(&owner)
+	}
+
+	header.SetCreationEpoch(opts.currentNeoFSEpoch)
+	currentVersion := version.Current()
+	header.SetVersion(&currentVersion)
+
 	res := &PayloadWriter{
-		stream:       x.w,
-		signer:       x.signer,
-		container:    x.cnr,
-		owner:        x.owner,
-		currentEpoch: x.opts.currentNeoFSEpoch,
-		sessionToken: x.sessionToken,
-		attributes:   attributes,
-		rootMeta:     newDynamicObjectMetadata(x.opts.withHomoChecksum),
-		childMeta:    newDynamicObjectMetadata(x.opts.withHomoChecksum),
-	}
-
-	res.buf.Grow(int(x.childPayloadSizeLimit()))
+		ctx:               ctx,
```

Comment:
Well, either we declare that `PutInit` context is for the whole operation or we remove this leaving it init-only and `Write()` undefined wrt context.

Reply:
We already have the first variant - we use this context for the whole operation. In my opinion, it is better than don't have any _context_ control of the operation

Reply:
But to have _some_ context to user per-request we can have a per-request timeout set in options for writer.

Reply:
So after some discussions we can tell that there is a conflict between `io.Writer` and `context.Context` in general. If we're to limit `ctx` to `InitPut` only then we'll get an unbound writer, so having a context for a whole PUT seems to be more useful since we definitely want to have `io.Writer` compatibility. It follows `net/http` and GRPC library implementation as well.

See #478 also.